src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldServer.java

```diff
@@ -1602,68 +1590,11 @@ public SpongeBlockSnapshot createSpongeBlockSnapshot(IBlockState state, IBlockSt
-            PhaseContext phaseContext = PhaseContext.start()
-                    .explosion()
-                    .addEntityCaptures()
-                    .addEntityDropCaptures()
-                    .addBlockCaptures();
-            if (entityIn != null) {
-                phaseContext.add(NamedCause.source(entityIn));
-            } else {
-                phaseContext.add(NamedCause.source(this));
-            }
-            final PhaseData currentPhaseData = CauseTracker.getInstance().getCurrentPhaseData();
-            currentPhaseData.state.getPhase().appendContextPreExplosion(phaseContext, currentPhaseData);
-            phaseContext.complete();
-            CauseTracker.getInstance().switchToPhase(GeneralPhase.State.EXPLOSION, phaseContext);
-        }
-        // Sponge End
-
         Explosion explosion = new Explosion((WorldServer) (Object) this, entityIn, x, y, z, strength, isFlaming, isSmoking);
-
-        // Sponge Start - More cause tracking
-        if (CauseTracker.ENABLED) {
-            try {
-                CauseTracker.getInstance().getCurrentContext().getCaptureExplosion().addExplosion(((org.spongepowered.api.world.explosion.Explosion) explosion));
-            } catch (Exception e) {
-                e.printStackTrace();
-            }
-        }
-        // Set up the pre event
-        final ExplosionEvent.Pre event = SpongeEventFactory.createExplosionEventPre(((IMixinExplosion) explosion).createCause(),
-                (org.spongepowered.api.world.explosion.Explosion) explosion, this);
-        if (SpongeImpl.postEvent(event)) {
-            this.processingExplosion = false;
-            if (CauseTracker.ENABLED) {
-                CauseTracker.getInstance().completePhase(GeneralPhase.State.EXPLOSION);
-            }
-            return explosion;
-        }
-        // Sponge End
-
-        explosion.doExplosionA();
-        explosion.doExplosionB(false);
-
-        if (!isSmoking) {
-            explosion.clearAffectedBlockPositions();
-        }
-
-        for (EntityPlayer entityplayer : this.playerEntities) {
-            if (entityplayer.getDistanceSq(x, y, z) < 4096.0D) {
-                ((EntityPlayerMP) entityplayer).connection.sendPacket(new SPacketExplosion(x, y, z, strength, explosion.getAffectedBlockPositions(),
-                        explosion.getPlayerKnockbackMap().get(entityplayer)));
-            }
-        }
-
-        // Sponge Start - end processing
-        this.processingExplosion = false;
-        if (CauseTracker.ENABLED) {
-            CauseTracker.getInstance().completePhase(GeneralPhase.State.EXPLOSION);
-        }
-        // Sponge End
+        // Sponge start - move this behavior to another method to allow precreated explosions to be used
+        // All of the original behavior is in triggerExplosion().
+        triggerExplosion((org.spongepowered.api.world.explosion.Explosion) explosion, Cause.of(NamedCause.source(entityIn == null ? this : entityIn)));
```

Comment:
I'd rather use the CauseTracker to define what `Cause` to use for the explosion source. Since the source passed in as `Entity` is usually an entity ticking, or in some cases a `TileEntity` passing null for explosions. You may also want to verify the explosion optimization (not really optimization, it's an optimization for clients on the server due to the unperformant explosion packets and what they entail) still functions as expected.